### ISSUE TITLE
Identifiers for Properties, Actions, Events

### ISSUE THREAD

**Issue by maximelefrancois86** [w3c/wot-thing-description#292]
Spec says: 

> Each instance of a Property, Action, and Event class MUST have an identifier that is unique within the context of a Thing Description document. See Section Representation Format for more details about JSON-LD 1.1 identifiers.

https://github.com/w3c/wot-thing-description/blob/807ca425dbce8868dbfd7f7d675d9ae99684512a/index.html#L487

- It is not said that the identifiers of  Property, Action, Event, MUST be JSON-LD 1.1 identifiers. What is meant by JSON-LD 1.1 identifier? How are these identifiers computed/defined.

- defined ->  it could be made more clear that the identifier is to be provided with `@id` key? maybe add this key explicitly in the Information Model?
- computed -> (in the RDF generation process, I heard @vcharpenay would like to compute unique identifiers concatenating the URI of the Thing with the JSON Pointer to the Prop/Act/Event) . How compatible is this with explicitly defined identifiers?

**Comment by mkovatsc** [comment]
I agree we need to more explicit explanation and definition for this. Aspects from the top of my head (maybe incomplete or incorrect):

* Properties, Actions, and Events MUST have a Thing-local identifier, whose value also must be restricted to a valid JS object attribute name (proper term and reference needed here), e.g., must not contain slashes, must start with alpha-num or underscore, etc.
* This Thing-local identifier is mapped to the `@id` of the respective node (which is done automatically through `@container: @id` and the expected future `@base: @id`.
* In `@id` form, it is a IRI.
* There must be an algorithm that applies the `@container: @id` and `@base: @id` processing, so a JSON-LD 1.0 processor can work with it.

**Comment by draggett** [comment]
Does the TD spec allow the same name to be used for properties, actions and events, or is the requirement only that you can't have two properties with the same name, and likewise for actions and events? 

What other constraints do we put on names for properties, actions and events?  I would suggest that we preclude names that start with "@" as this would otherwise make TDs harder to read.

I would assume that each property has exactly one name, but may have multiple descriptions when the developer wants to document a thing in more than one human language. I think we are waiting on guidance for the mechanism for doing that.

Note the terminological question about names and identifiers.  I've used name in the sense of an identifier above, but we could allow developer's to provide localised identifiers as metadata once we have guidance as how to do so.

**Comment by vcharpenay** [comment]
I've started specifying an algorithm to turn a TD into JSON-LD 1.0. See https://wot-td-ontology.github.io/wot-thing-description/ontology/#algorithm. The algorithm takes a _pointer_ as input, which must be the TD `id` field.

**Comment by mmccool** [comment]
Dave, Note that the issue is that "field names" need to be unique within a JSON-LD document.   So, no, you can't have a property and an action with the same name, or even a security definition and a property with the same name.   This is kind of annoying, but not the end of the world.   I think at this point we should just figure out how to clearly state this limitation in the spec, not try to fix it.   In a later revision of the spec, once JSON-LD 1.1 is out, maybe they will have a fix and this limitation can be lifted (lifting it will still preserve backward compatibility; old TDs with unique names will, of course, still work).

**Comment by draggett** [comment]
Contrariwise, We could consider it an advantage not being able to have a property and an action with the same name since that is also the case for languages like JavaScript.

**Comment by mkovatsc** [comment]
I am surprised this is still open -- I expected that the identifiers / "field names" to be unique across the whole Thing (no Property and Action with the same identifier/name).

This is required to make `@container: @id` for JSON-LD 1.1 work and it also simplifies (Scripting) API issues.

I am not sure what Victor's standpoint in https://github.com/w3c/wot-thing-description/issues/292#issuecomment-441114183 is.

The current TD spec has a statement that agrees: 

> Each instance of a Property, Action, and Event class MUST have an identifier that is unique within the context of a Thing Description document. 

Maybe it should be:

> Each instance of a Property, Action, and Event class MUST have an identifier that MUST be unique within the context of a Thing Description document.

**Comment by egekorkan** [comment]
Indeed this is an open issue. To clarify, do we mean:

1.  All interaction patterns must have unique names, so an action and a property cannot have the same name (What I was understanding until now)
2. In addition to 1., there cannot be a security definition with same name as an action OR even a property called `support` if in the Thing level there is the support key OR a property key of an `ObjectSchema` with the same name as an action. (What I understand by @mmccool 's comment above)
3. In addition to 2., an interaction pattern cannot have a name that belongs to the TD vocabulary, such as `href`, `anyOf`, etc. even if this "reserved" key doesn't appear in the given TD. (Not mentioned by anyone yet)

**Comment by mmccool** [comment]
Everything Ege says is correct but what I must say is that the current assertion is not clear enough.  The real technical requirement is that all object keys that map to ids need to be unique.  The current statement could still be misread as (for example) property names only need to be unique among the set of property names in the Thing, which is wrong.

We need to more carefully word this assertion, and I think this issue should stay open until we do.  BTW, please don’t combine multiple assertions into one sentence, it complicates testing ;).   We already had to break up several...

**Comment by mmccool** [comment]
Correction to “I agree with what Ege said”: there’s one point where what he says is not accurate, or at least I hope so.

I was assuming that we *can* have a property (for example) with the same name as a built-in term (eg “support”).  What really matters is that the URIs are distinct, and I hope we can arrange for this to be true. (If not, we definitely need yet another assertion...)

It MAY also be possible to generate different URIs for (say) properties and actions with the same name, or (more relevantly, perhaps) securityDefinitions and properties with the same name.  If we can’t, and these are in the same namespace, we need to (more) clearly state that.

**Comment by mkovatsc** [comment]
> We need to more carefully word this assertion, and I think this issue should stay open until we do.

Okay. I think this issue confirmed that in addition to the "MUST be unique within the context of a Thing Description document", we should elaborating by stating that there cannot be a Property and Action or Event having the same identifier, nor a security definition having the same identifier as a Property, Action, or Event.

> BTW, please don’t combine multiple assertions into one sentence, it complicates testing ;)

Sure, make two sentences. My main point was that besides the "having an identifier" assertion, we also need a "uniqueness" assertion.

> even a property called `support` if in the Thing level there is the support key

> interaction pattern cannot have a name that belongs to the TD vocabulary

These would be the same limitation, however, this is wrong. TD terms are Predicates, not `@id`s. If you experience conflicts, maybe this is a limitation of the JSON Schema checker?

I am not fully sure about `oneOf` right now. I think they should become blank nodes, which is actually fine.

> a property key of an `ObjectSchema` with the same name as an action. (What I understand by @mmccool 's comment above)

Here the plan was to change the `@base` of the local Property context, so that sub-properties basically end up in a URI path structure. This is still something the JSON-LD 1.1 WG needs to support, but I think we had a good use case ("idiomatic JSON").

@vcharpenay Are you taking this into account in the transformation? BTW is there a link anywhere to that document? I don't even know the title despite it being an important building block...

> It MAY also be possible to generate different URIs for (say) properties and actions

Yes, altering `@base` might work, but it would not be a generic mechanism like above, as basically new intermediate nodes are introduced between the Thing node and the Property, Action, and Event nodes.

Overall, also having looked at the Scripting API issues, I think it is rather a feature to have Thing-wide uniqueness of the different user-defined identifiers for "first-class citizens of a Thing" (Properties, Actions, Events, Security Definitions -- not sub-properties or input/output).

**Comment by egekorkan** [comment]
> If you experience conflicts, maybe this is a limitation of the JSON Schema checker?

I don't have any problem at the moment but it was just to clarify. 

> These would be the same limitation,

I see that it sounds like the same limitation but with the second clause, I meant `"amazingMetaData":"wow"` at the root of the TD and having an action called `amazingMetaData`. From your explanation, I see that this is a wrong limitation as well.

> TD terms are Predicates, not `@id`s

However, this is something that I don't really understand.

**Comment by egekorkan** [comment]
To clarify for testing, please comment if the following is false:
- When combining the property, action, event and securityDefinition keys into an array, there should not be any duplicates

**Comment by vcharpenay** [comment]
> @vcharpenay Are you taking this into account in the transformation? BTW is there a link anywhere to that document? I don't even know the title despite it being an important building block...

@mkovatsc there was a decision to include the algorithm in a non-normative part of the TD spec: see [Implementation note](https://w3c.github.io/wot-thing-description/#Implementation-Note-section) section.

> Each instance of a Property, Action, and Event class MUST have an identifier that is unique within the context of a Thing Description document.

In the transformed TD ouput by the algorithm, the assertion above do not hold anymore : the whole JSON pointer of every definition is part of the `@id`. There is no need to enforce that property, action and event names are unique within a TD. As @mmccool said, only the generated `@id` must be unique.

**Comment by mmccool** [comment]
As a candidate text for replacing the current assertion, I think this is reasonable, but there may be some confusion from readers who may think we mean "array" as an actual type in the JSON file.   I would use a slightly more abstract wording with the same meaning, and with RFC2119 MUST:

- When combining the names of properties, actions, events, and securityDefinitions into a single collection, there MUST not be any duplicates.

I would also add a non-normative explanatory sentence right afterwards:

- For example, two properties cannot have the same name; an action and a property cannot have the same name; and a securityDefinition cannot have the same name as any property, event, or action.

Here I use "name" rather than "key".   Actually, I forgot again what we were supposed to be using for this element of the JSON file, so before I would edit the doc I would look up what term we decided to use for this and modify the above appropriately.

**Comment by mmccool** [comment]
Ah, I overlooked @vcharpenay comment that this issue is resolved by putting the JSON pointer into the generated @id, so this assertion is no longer needed.  That is excellent!   I will therefore create a PR to remove this assertion (rather than modifying it as I suggested above).

**Comment by mmccool** [comment]
I created a PR to remove td-unique-identifiers, which in conjuction with the update by @vcharpenay to automatically generated unique @ids should resolve this issue.  I note there is still an ambiguous use of "unique" in various assertions which should however be resolved (not in this PR though...) with an appropriate definition and use of "Map of ..." when defining the types of those vocabulary terms that use key-value mappings rather than simple objects.

**Comment by mmccool** [comment]
This issue can be closed now.  It has been resolved so that identifiers only need to be unique within an object's scope.

**Comment by zolkis** [comment]
My own opinion is that for any scripting use cases (in any language) it would be an advantage if all interaction identifiers were unique within a Thing (Description). That means we should not allow in a TD to have a property with the same name as an action: they are both interactions.
So I wish we had a Thing (Description) as a single namespace.

**Comment by vcharpenay** [comment]
here, names are for indexing purposes. When deserializing JSON, all objects will have some (hidden) identifier assigned by the JS runtime. If a global identifier is required for individual affordances, [RDF  normalization](http://json-ld.github.io/normalization/spec/) could be applied. All relevant TD objects would then have a hash URI as `@id`.

**Comment by zolkis** [comment]
> here, names are for indexing purposes. 

Indexing in what namespace?

> When deserializing JSON, all objects will have some (hidden) identifier assigned by the JS runtime. If a global identifier is required for individual affordances, RDF  normalization could be applied. All relevant TD objects would then have a hash URI as @id.

Identifiers in scripting runtime are definitely not hidden. We must expose object-unique key names.
The question is who decides whether normalization is done or not (application or scripting implementation or TD writer), and what can a scripting implementation "normalize" in scripting terms in order to make developer-friendly and intuitive/recognisable object key names out of names and/or `@id`s. Do we have an algorithm for that? If not, we must specify one, since without that there is no point talking about any Scripting API. This issue cannot be closed yet without clarifying this.

This makes the point that over-generalization actually kills a use case. What would it cost us or TD writers to require that for developers' sake, the interaction affordance names should be [valid scripting object names](http://www.ecma-international.org/ecma-262/9.0/index.html#sec-object-type)?

**Comment by zolkis** [comment]
@danielpeintner do you have an insight on the above?

**Comment by vcharpenay** [comment]
> This makes the point that over-generalization actually kills a use case. What would it cost us or TD writers to require that for developers' sake, the interaction affordance names should be valid scripting object names?

well, in ECMAScript, property keys can be any ECMAScript string. In the JSON TD serialization, the restriction is the same, since JSON is based on ECMAScript.

**Comment by vcharpenay** [comment]
At this point, the discussion significantly drifted away from the original point. I'll close the issue. @zolkis please open a new one if you think it is necessary.

**Comment by zolkis** [comment]
Not really needed, I asked @danielpeintner to explain this in the Scripting call.
